hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/DefaultJobPartitionProvider.java

```diff
@@ -0,0 +1,125 @@
+	 * @param theRequestDetails the requestDetails
+	 * @param theOperation the operation being run which corresponds to the job
+	 * @return the list of {@link PartitionedUrl}
+	 */
+	public List<RequestPartitionId> getPartitions(RequestDetails theRequestDetails, String theOperation) {
+		RequestPartitionId partitionId = myRequestPartitionHelper.determineReadPartitionForRequestForServerOperation(
+				theRequestDetails, theOperation);
+		return List.of(partitionId);
+	}
+
+	/**
+	 * Provides the list of {@link PartitionedUrl} to run job steps against, based on the request that initiates the job
+	 * and the urls that it's configured with.
+	 * @param theRequestDetails the requestDetails
+	 * @param theUrls the urls to run the job against
+	 * @return the list of {@link PartitionedUrl}
+	 */
+	@Override
+	public List<PartitionedUrl> getPartitionedUrls(RequestDetails theRequestDetails, @NotNull List<String> theUrls) {
+		List<String> urls = theUrls;
+
+		// if the url list is empty, use all the supported resource types to build the url list
+		if (theUrls.isEmpty()) {
+			urls = myFhirContext.getResourceTypes().stream()
+					.map(resourceType -> resourceType + "?")
+					.collect(Collectors.toList());
+		}
+
+		// determine the partition associated with each of the urls
+		List<PartitionedUrl> partitionedUrls = new ArrayList<>();
+		for (String s : urls) {
+			ResourceSearch resourceSearch = myMatchUrlService.getResourceSearch(s);
+			RequestPartitionId partitionId = myRequestPartitionHelper.determineReadPartitionForRequestForSearchType(
+					theRequestDetails, resourceSearch.getResourceName(), resourceSearch.getSearchParameterMap());
+			partitionedUrls.add(new PartitionedUrl().setUrl(s).setRequestPartitionId(partitionId));
+		}
+
+		// handle (bulk) system operations that are typically configured with RequestPartitionId.allPartitions()
+		// populate the actual list of all partitions, if that is supported
+		Set<RequestPartitionId> allPartitions = new LinkedHashSet<>(getAllPartitions());
+		if (!allPartitions.isEmpty()) {
+			allPartitions.add(RequestPartitionId.defaultPartition());
+		}
+		List<PartitionedUrl> retVal = new ArrayList<>();
+		for (PartitionedUrl partitionedUrl : partitionedUrls) {
+			String url = partitionedUrl.getUrl();
+			RequestPartitionId partitionId = partitionedUrl.getRequestPartitionId();
+			if (partitionId != null && partitionId.isAllPartitions() && !allPartitions.isEmpty()) {
+				List<PartitionedUrl> megaScalePartitionedUrls = allPartitions.stream()
+						.map(p -> (new PartitionedUrl().setUrl(url).setRequestPartitionId(p)))
+						.collect(Collectors.toList());
+				retVal.addAll(megaScalePartitionedUrls);
+			} else {
+				retVal.add(partitionedUrl);
+			}
+		}
+
+		return retVal;
+	}
+
+	public List<RequestPartitionId> getAllPartitions() {
+		// likely allPartitions is not supported in this path
+		return List.of();
```

Comment:
I'm not sure we still need the JPA version of this class.

```suggestion
		return List.of(RequestPartitionId.allPartitions());
```